### DISCUSSION TITLE
What is the best practice to setup service network inside the DAG?

### DISCUSSION THREAD

**Discussion by ufnalbartosz** [apache/airflow#27218] [Q&A] [answered]
I'm trying to make a pipeline where I have one forever running webservice (Jena Fuseki) and a second that would make queries to the Fuseki (Robot Framework).
So far I've used the `KubernetesPodOperator`  and the `hostnetwork=True` parameter. However, I have trouble setting the correct IP address inside these services because the metadata `host_ip` changes frequently.

I also don't quite understand the proper use of the `airflow.utils.net.get_host_ip_address` because when I used it, it returned a different `host_ip` than the metadata `host_ip`.

```python

HOST_IP = airflow.utils.net.get_host_ip_address()

fuseki_port = k8s.V1ContainerPort(name="fuseki-port", container_port=3030)

fuseki = KubernetesPodOperator(
    name="k8s-start-fuseki",
    task_id="start_fuseki",
    namespace=DEFAULT_NAMESPACE,
    image=FUSEKI_IMAGE,
    is_delete_operator_pod=False,
    in_cluster=True,
    dag=dag,
    hostnetwork=True,
    ports=[fuseki_port],
    image_pull_policy="Always",
    get_logs=True,
    env_vars={"HOSTNETWORK_IP": HOST_IP},
)

robot_port = k8s.V1ContainerPort(name="robot-port", container_port=8000)

robot = KubernetesPodOperator(
    name="k8s-start-robot",
    task_id="start_robot",
    namespace=DEFAULT_NAMESPACE,
    image=ROBOT_FRAMEWORK_IMAGE,
    is_delete_operator_pod=False,
    in_cluster=True,
    dag=dag,
    hostnetwork=True,
    ports=[fuseki_port, robot_port],
    image_pull_policy="Always",
    get_logs=True,
    env_vars={"HOSTNETWORK_IP": HOST_IP},

)

``` 

With this example I'm getting `HTTPSConnectionPool(host='100.83.X.X', port=3030): Max retries exceeded` inside the Robot Framework's KubernetesPodOperator.

Is there any better way of setting the correct IP address? Maybe even is there a better way to spin two web services that can communicate between themselves inside the DAG?

**Comment by potiuk** [comment, marked as answer]
There are no practices for setting webserver inside your DAG execution - I thnik what you are trying to do is definitely not what Airflow is intended to work. With airflow's model setting "internal" IP is very much dependent on the deployment you have and my advice is - like with any other services which Airflow can start/stop - make sure they are accessible over public address (migt be inside your VPN)  and having a DNS name and connect to it. This is the best practice I can recommend.

**Reply by ufnalbartosz** [reply]
Thanks for your answer! But is there any way to create any Cluster Networking as it would be in K8s Pod-to-Pod? Between two or more `KubernetesPodOperators`?

**Reply by ufnalbartosz** [reply]
Would a correct use of the [`pod_mutation_hook`](https://airflow.apache.org/docs/apache-airflow/stable/kubernetes.html#pod-mutation-hook) be a good choice in my scenario?

**Reply by potiuk** [reply]
I have no idea what you really want to do :) so it is hard to say, but pod_mutation_hook can modify definition of the hook dynamically (however if you do not need to modify it dynamically at the moment of creation you can do all the same in pod_template - you can define your POD in any way you want. As long as your POD definition can be uploaded to K8S via API and do whatever you want and it will be always the same in all instances of the same task, you can do it in the pod_template file. The pod_mutation_hook is used only if you want to verify or update the definition dynamically at execution time. But maybe I do not understand your scenario and what you are trying to do - or maybe you do not understand how airflow Kubernetes Pod Operator or Kubernetes executor works. Each of them simply at the end create a POD, wait for it to happen, monitor its execution and streams logs and finally delets the pod when done. Each of the pods is independent on each other.  Maybe read more about ti in the docs / watch some videos if unsure? 

I think you are effectively trying to run some communication between those two pods - but this is far beyond what Airflow does - it assumes those tasks are independent and if you want to facilitate any communication between them you need to likely do more than just create PODS (likely you need to have some service definitions and the like) - but this is nothing that airflow handles out of the box and you need ot do it in Kubernetes way, not airflow way.

Airflow just submit PODS which you can define. That's about it.
